### PR TITLE
Add Riemannian Potato Field

### PR DESCRIPTION
**Pull request by qbarthelemy**
This PR:
- adds the class `PotatoField`,
- adds an animated example in the gallery, on artifact detection by  Riemannian Potato Field , compared with  Riemannian Potato.

Q. Barthélemy, L. Mayaud, D. Ojeda, M. Congedo, "The Riemannian potato field: a tool for online signal quality index of EEG", IEEE TNSRE, 2019, [link to pdf](https://hal.archives-ouvertes.fr/hal-02015909/document).

### CODE REVIEW

File: examples/artifacts/plot_detect_riemannian_potato_field_EEG.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+train_set = range(t)
+rp.fit(rp_covs[train_set])
+
+
+###############################################################################
+# Riemannian potato field
+# -----------------------
+#
+# Riemannian potato field (RPF) [1]_ combines several potatoes of low dimension
+# , each one being designed to capture specific artifact typically affecting
```
Comment (agramfort) [inline examples/artifacts/plot_detect_riemannian_potato_field_EEG.py:120]: breaking , here is weird

File: examples/artifacts/plot_detect_riemannian_potato_field_EEG.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
+"""
+===============================================================================
+Online Artifact Detection with Riemannian Potato Field
+===============================================================================
+
+Example of Riemannian Potato Field [1]_ applied on EEG time-series to detect
```
Comment (agramfort) [inline examples/artifacts/plot_detect_riemannian_potato_field_EEG.py:6]:
```suggestion
Example of Riemannian Potato Field (RPF) [1]_ applied on EEG time-series to detect
```
as you use the acronym below

File: examples/artifacts/plot_detect_riemannian_potato_field_EEG.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+rpf.fit([c[train_set] for c in rpf_covs])
+
+
+###############################################################################
+# Online Artifact Detection with Potatoes
+# ---------------------------------------
+#
+# Detect artifacts/outliers on test set, with an animation to imitate an online
+# acquisition, processing and artifact detection of EEG time-series.
+# Remak that all these potatoes are semi-dynamic: they are updated when EEG is
```
Comment (agramfort) [inline examples/artifacts/plot_detect_riemannian_potato_field_EEG.py:170]:
```suggestion
# Remark that all these potatoes are semi-dynamic: they are updated when EEG is
```

File: examples/artifacts/plot_detect_riemannian_potato_field_EEG.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+    for c in range(len(pl2)):
+        pl2[c].set_data(covs_t, covs_z[c])
+        pl2[c].axes.set_xlim(covs_t[0] - 0.1, covs_t[-1])
+    pl3[0].set_data(covs_t, covs_p)
+    return pl, pl2, pl3
+
+
+###############################################################################
+
+# Plot online detection (a dynamic display is required)
```
Comment (agramfort) [inline examples/artifacts/plot_detect_riemannian_potato_field_EEG.py:264]:
```suggestion
###############################################################################
# Plot online detection (a dynamic display is required)
```

File: pyriemann/clustering.py [hunk 5]
```diff
@@ -483,11 +489,231 @@ def _check_labels(self, X, y):
...
 
     def _get_proba(self, z):
-        """get right-tailed proba from z-score."""
+        """Get right-tailed proba from z-score."""
         proba = 1 - norm.cdf(z)
         return proba
+
+
+class PotatoField(BaseEstimator, TransformerMixin, ClassifierMixin):
+
```
Comment (agramfort) [inline pyriemann/clustering.py:503]:
```suggestion
class PotatoField(BaseEstimator, TransformerMixin, ClassifierMixin):
```

File: pyriemann/clustering.py [hunk 6]
```diff
@@ -483,11 +489,231 @@ def _check_labels(self, X, y):
...
+        outlier removal to obtain reliable potatoes.
+
+        Parameters
+        ----------
+        X : list of n_potatoes ndarray of shape (n_trials, n_channels,
+                n_channels), with same n_trials but different n_channels
+            list of ndarray of SPD matrices.
+        y : ndarray, shape (n_trials,) | None (default None)
+            Labels corresponding to each trial. If None, all trials are
+            considered as clean.
```
Comment (agramfort) [inline pyriemann/clustering.py:565]: otherwise what is considered as clean?

File: pyriemann/clustering.py [hunk 7]
```diff
@@ -483,11 +489,231 @@ def _check_labels(self, X, y):
...
+
+        Parameters
+        ----------
+        X : list of n_potatoes ndarray of shape (n_trials, n_channels,
+                n_channels), with same n_trials but different n_channels
+            list of ndarray of SPD matrices.
+
+        Returns
+        -------
+        z : ndarray, shape (n_potatoes, n_trials)
```
Comment (agramfort) [inline pyriemann/clustering.py:642]:
```suggestion
        z : ndarray, shape (n_trials, n_potatoes)
```
would be more natural

File: tests/test_clustering.py [hunk 8]
```diff
@@ -221,3 +260,57 @@ def test_Potato_specific_labels(get_covmats):
     assert_array_equal(np.unique(pt.predict(covmats)), [2, 7])
     # fit with custom positive label
     pt.fit(covmats, y=[2] * n_trials)
+
+
+def test_PotatoField_fit(get_covmats):
```
Comment (agramfort) [inline tests/test_clustering.py:265]:
```suggestion
def test_potato_field_fit(get_covmats):
```

File: tests/test_clustering.py [hunk 9]
```diff
@@ -221,3 +260,57 @@ def test_Potato_specific_labels(get_covmats):
...
+    with pytest.raises(ValueError):  # p_threshold out of bounds
+        PotatoField(p_threshold=1).fit(covmats)
+    pf = PotatoField(n_potatoes=n_potatoes)
+    with pytest.raises(ValueError):  # n_potatoes not equal to input length
+        pf.fit([covmats1, covmats1, covmats2])
+    with pytest.raises(ValueError):  # n_trials not equal
+        pf.fit([covmats1, covmats2[:1]])
+
+
+def test_PotatoField_partialfit(get_covmats):
```
Comment (agramfort) [inline tests/test_clustering.py:283]:
```suggestion
def test_potato_field_partial_fit(get_covmats):
```

File: tests/test_clustering.py [hunk 10]
```diff
@@ -221,3 +260,57 @@ def test_Potato_specific_labels(get_covmats):
...
+    covmats2 = get_covmats(n_trials, n_channels + 1)
+    covmats = [covmats1, covmats2]
+    pf = PotatoField(n_potatoes=n_potatoes).fit(covmats)
+    with pytest.raises(ValueError):  # n_potatoes not equal to input length
+        pf.partial_fit([covmats1, covmats1, covmats2])
+    with pytest.raises(ValueError):  # n_trials not equal
+        pf.partial_fit([covmats1, covmats2[:1]])
+
+
+def test_PotatoField_transform(get_covmats):
```
Comment (agramfort) [inline tests/test_clustering.py:295]:
```suggestion
def test_potato_field_transform(get_covmats):
```

File: tests/test_clustering.py [hunk 11]
```diff
@@ -221,3 +260,57 @@ def test_Potato_specific_labels(get_covmats):
...
+    covmats2 = get_covmats(n_trials, n_channels + 1)
+    covmats = [covmats1, covmats2]
+    pf = PotatoField(n_potatoes=n_potatoes).fit(covmats)
+    with pytest.raises(ValueError):  # n_potatoes not equal to input length
+        pf.transform([covmats1, covmats1, covmats2])
+    with pytest.raises(ValueError):  # n_trials not equal
+        pf.transform([covmats1, covmats2[:1]])
+
+
+def test_PotatoField_predictproba(get_covmats):
```
Comment (agramfort) [inline tests/test_clustering.py:307]:
```suggestion
def test_potato_field_predict_proba(get_covmats):
```

**Review by sylvchev — APPROVED** [review]

File: pyriemann/clustering.py (hunk 6 above)
Comment (qbarthelemy) [reply pyriemann/clustering.py:565]: Trials having labels `y` equal to `pos_label` are considered as clean (explained in the doc of Potato and PotatoField parameters).

Comment (agramfort) [reply pyriemann/clustering.py:565]: can you clarify this docstring accordingly? 🙏

**Review by agramfort — APPROVED** [review]

File: examples/artifacts/plot_detect_riemannian_potato_field_EEG.py [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+
+
+###############################################################################
+# Riemannian potato field
+# -----------------------
+#
+# Riemannian potato field (RPF) [1]_ combines several potatoes of low
+# dimension, each one being designed to capture specific artifact typically
+# affecting specific spatial areas (ie, subsets of channels) and/or specific
+# frequency bands.
```
Comment (plcrodrigues) [inline examples/artifacts/plot_detect_riemannian_potato_field_EEG.py:123]:
```suggestion
# Riemannian potato field (RPF) [1]_ combines several potatoes of low
# dimensionality, each one designed to capture a different kind of artifact typically
# affecting some specific spatial area (i.e. subsets of channels) and/or specific
# frequency bands.
```

File: pyriemann/clustering.py [hunk 13]
```diff
@@ -483,11 +491,233 @@ def _check_labels(self, X, y):
...
+
+    def fit(self, X, y=None):
+        """Fit the potato field from covariance matrices, with iterative
+        outlier removal to obtain reliable potatoes.
+
+        Parameters
+        ----------
+        X : list of n_potatoes ndarray of shape (n_trials, n_channels,
+                n_channels), with same n_trials but different n_channels
+            list of ndarray of SPD matrices.
```
Comment (plcrodrigues) [inline pyriemann/clustering.py:564]:
```suggestion
        X : list of n_potatoes ndarrays of shape (n_trials, n_channels,
            n_channels) with same n_trials but potentially different n_channels
            List of ndarrays of SPD matrices, each corresponding to a different 
            subset of EEG channels and/or filtering with a specific frequency band.
```

File: pyriemann/clustering.py [hunk 14]
```diff
@@ -483,11 +491,233 @@ def _check_labels(self, X, y):
...
+        """Partially fit the potato field from covariance matrices.
+
+        This partial fit can be used to update dynamic or semi-dymanic online
+        potatoes with clean EEG [1]_.
+
+        Parameters
+        ----------
+        X : list of n_potatoes ndarray of shape (n_trials, n_channels,
+                n_channels), with same n_trials but different n_channels
+            list of ndarray of SPD matrices.
```
Comment (plcrodrigues) [inline pyriemann/clustering.py:609]: Same as above.

File: pyriemann/clustering.py [hunk 15]
```diff
@@ -483,11 +491,233 @@ def _check_labels(self, X, y):
...
+            corresponds to a clean (resp. artifact) trial. If None, all trials
+            are considered as clean.
+        alpha : float (default 0.1)
+            Update rate in [0, 1] for the centroid, and mean and standard
+            deviation of log-distances: 0 for no update, 1 for full update.
+
+        Returns
+        -------
+        self : Potato instance
+            The Potato instance.
```
Comment (plcrodrigues) [inline pyriemann/clustering.py:621]:
```suggestion
        self : PotatoField instance
            The Potato Field instance.
```

File: pyriemann/clustering.py [hunk 16]
```diff
@@ -483,11 +491,233 @@ def _check_labels(self, X, y):
...
+        z = np.zeros((n_trials, self.n_potatoes))
+        for i in range(self.n_potatoes):
+            if X[i].shape[0] != n_trials:
+                raise ValueError(
+                    'Unequal n_trials between ndarray of X. Should be %d but '
+                    'got %d.' % (n_trials, X[i].shape[0]))
+            z[:, i] = self._potatoes[i].transform(X[i])
+        return z
+
+    def predict(self, X, alpha=0):
```
Comment (plcrodrigues) [inline pyriemann/clustering.py:661]:
```suggestion
    def predict(self, X):
```

File: pyriemann/clustering.py [hunk 17]
```diff
@@ -483,11 +491,233 @@ def _check_labels(self, X, y):
...
+    Parameters
+    ----------
+    n_potatoes : int (default 1)
+        Number of potatoes in the field.
+    p_threshold : float (default 0.01)
+        Threshold on probability to being clean, in (0, 1).
+    metric : string (default 'riemann')
+        The type of metric used for centroid and distance estimation.
+    z_threshold : float (default 3)
+        Threshold on z-score of distance to the centroid.
```
Comment (plcrodrigues) [inline pyriemann/clustering.py:522]:
```suggestion
    z_threshold : float (default 3)
        Threshold on z-score of distance to reject artifacts. It is the number
        of standard deviations from the mean of distances to the centroid.    
```

File: pyriemann/clustering.py [hunk 18]
```diff
@@ -483,11 +491,233 @@ def _check_labels(self, X, y):
...
+
+    References
+    ----------
+    .. [1] Q. Barthélemy, L. Mayaud, D. Ojeda, M. Congedo, "The Riemannian
+        potato field: a tool for online signal quality index of EEG", IEEE
+        TNSRE, 2019.
+    """
+
+    def __init__(self, n_potatoes=1, p_threshold=0.01, metric='riemann',
+                 z_threshold=3, n_iter_max=10, pos_label=1, neg_label=0):
```
Comment (plcrodrigues) [inline pyriemann/clustering.py:546]:
```suggestion
    def __init__(self, n_potatoes=1, p_threshold=0.01, z_threshold=3,
                 metric='riemann', n_iter_max=10, pos_label=1, neg_label=0):
```

File: pyriemann/clustering.py [hunk 19]
```diff
@@ -483,11 +491,233 @@ def _check_labels(self, X, y):
...
+    artifact in EEG signals. The algorithm combines several potatoes of low
+    dimension, each one being designed to capture specific artifact typically
+    affecting specific subsets of channels and/or specific frequency bands.
+
+    Parameters
+    ----------
+    n_potatoes : int (default 1)
+        Number of potatoes in the field.
+    p_threshold : float (default 0.01)
+        Threshold on probability to being clean, in (0, 1).
```
Comment (plcrodrigues) [inline pyriemann/clustering.py:518]: Can you be more precise?

Comment (qbarthelemy) [reply pyriemann/clustering.py:518]: Ok, I complete the doc.

**Review by plcrodrigues — APPROVED** [review]

**Comment by nbara** [comment]
Hi @qbarthelemy, one of the figures is empty in the doc : 

https://pyriemann.readthedocs.io/en/latest/auto_examples/artifacts/plot_detect_riemannian_potato_field_EEG.html#sphx-glr-auto-examples-artifacts-plot-detect-riemannian-potato-field-eeg-py

Is that normal?

**Comment by qbarthelemy** [comment]
Yes, it's normal: since it is an online processing, a dynamic display is required.
In you want to see the animation, run the example from a shell.

But I have tried to plot the animation in the html documentation, but I have not found the way to do not display it twice.
https://github.com/pyRiemann/pyRiemann/pull/133#issuecomment-890333261
Your ideas are welcome!

**Pull request merged**